src/prepare/compile.rs

```diff
@@ -0,0 +1,1239 @@
+            F64ConvertUI32 => {
+                context.step(instruction)?;
+                self.sink.emit(isa::InstructionInternal::F64ConvertUI32);
+            }
+            F64ConvertSI64 => {
+                context.step(instruction)?;
+                self.sink.emit(isa::InstructionInternal::F64ConvertSI64);
+            }
+            F64ConvertUI64 => {
+                context.step(instruction)?;
+                self.sink.emit(isa::InstructionInternal::F64ConvertUI64);
+            }
+            F64PromoteF32 => {
+                context.step(instruction)?;
+                self.sink.emit(isa::InstructionInternal::F64PromoteF32);
+            }
+
+            I32ReinterpretF32 => {
+                context.step(instruction)?;
+                self.sink.emit(isa::InstructionInternal::I32ReinterpretF32);
+            }
+            I64ReinterpretF64 => {
+                context.step(instruction)?;
+                self.sink.emit(isa::InstructionInternal::I64ReinterpretF64);
+            }
+            F32ReinterpretI32 => {
+                context.step(instruction)?;
+                self.sink.emit(isa::InstructionInternal::F32ReinterpretI32);
+            }
+            F64ReinterpretI64 => {
+                context.step(instruction)?;
+                self.sink.emit(isa::InstructionInternal::F64ReinterpretI64);
+            }
+            _ => {
+                context.step(instruction)?;
+            }
+        };
+
+        assert_eq!(self.label_stack.len(), context.frame_stack.len(),);
+
+        Ok(())
+    }
+}
+
+fn compute_drop_keep(
+    in_stack_polymorphic_state: bool,
+    started_with: StartedWith,
+    block_type: BlockType,
+    actual_value_stack_height: usize,
+    start_value_stack_height: usize,
+) -> Result<isa::DropKeep, Error> {
+    // Find out how many values we need to keep (copy to the new stack location after the drop).
+    let keep: isa::Keep = match (started_with, block_type) {
+        // A loop doesn't take a value upon a branch. It can return value
+        // only via reaching it's closing `End` operator.
+        (StartedWith::Loop, _) => isa::Keep::None,
+
+        (_, BlockType::Value(_)) => isa::Keep::Single,
+        (_, BlockType::NoResult) => isa::Keep::None,
+    };
+
+    // Find out how many values we need to discard.
+    let drop = if in_stack_polymorphic_state {
+        // Polymorphic stack is a weird state. Fortunately, it is always about the code that
+        // will not be executed, so we don't bother and return 0 here.
+        0
+    } else {
+        if actual_value_stack_height < start_value_stack_height {
+            return Err(Error(format!(
+                "Stack underflow detected: value stack height ({}) is lower than minimum stack len ({})",
+                actual_value_stack_height,
+                start_value_stack_height,
+            )));
+        }
+        if (actual_value_stack_height as u32 - start_value_stack_height as u32) < keep as u32 {
```

Comment:
Wouldn't exhaustive match for keep be much more clear of what's going on and will avoid all this casts?

```rust
match keep {
   isa::Keep::Single => if actual_value_stack_height + 1 < start_value_stack_height { ... },
   isa::Keep::Single = { ... },
   isa::Keep::None =>  { ... }
}
```